### PR TITLE
add nova prefilter for blazar

### PR DESCRIPTION
**Pull request by msherman64**
We've been having issues with the blazar filter seing stale "host_state", and acting on incorrect info about aggregate membership. Our use of host_state is nearly worst-case, since we use only a single ironic-conductor, the host_state contains information about every single node.

Most of blazar's behavior could be implemented by taking advantage of the existing Nova prefilters, which read from the DB directly rather than parsing the host_state. This filtering takes place prior to making the request to placement, which besides not being subject to stale data, should significantly decrease the request sizes between nova and placement.

What we need the filters to do is listed below, along with ways we could adapt blazar to perform them with no code changes in nova.

## Restrict instance launches to specified reservation

leverage `map_az_to_placement_aggregate` prefilter

in nova-scheduler.conf, set `query_placement_for_availability_zone=True`
in launch request, pass `availability_zone` instead of `reservation` hint.

## Ensure tenants own the requested reservation

leverage `require_tenant_aggregate` prefilter

in nova-scheduler.conf, set: `require_tenant_aggregate=True`
when blazar creates and aggregate, set `filter_tenant_id` = project_id

## Prevent instance launches without a reservation

leverage `isolate_aggregates` prefilter

in nova-scheduler.conf, set `scheduler.enable_isolated_aggregate_filtering=True`

On freepool, set `trait:launch_without_reservation=required`

Alternately, this PR attaches a composite "blazar filter" which merely implements the above, using the logic from the mentioned filters.

This adds the following two config items for nova's "[scheduler]" section

```
blazar_reservation_required = True # whether the prefilter should deny all requests with no reservation hint
use_blazar_reservation_prefilter = True # whether to enable this prefilter
```

### CODE REVIEW

**Comment by msherman64** [comment]
Even if we choose not to implement any filtering at this stage, this is where we could add a debug hook to report on the difference between the following "no hosts available" failure cases:

* user did not pass a reservation hint
* reservation hint did not correspond to an aggregate
* aggregate exists, but user does not have permission (or project ownership, currently)
* aggregate exists, but no hosts in it
* aggregate exists, and there are hosts in it, but those hosts were filtered out by nova for some other reason (there was an instance on it, the node was down, etc.)

**Review by Mark-Powers — APPROVED** [review]
This looks good. I would prefer a solution where we are not duplicating code, but it's probaby OK in this instance. We would also disable the blazar-nova filter with this change I am assuming?

File: nova/scheduler/request_filter.py [hunk 1]
```diff
@@ -134,6 +133,68 @@ def require_tenant_aggregate(ctxt, request_spec):
...
+    LOG.debug(f"entered blazar prefilter with request_spec {request_spec}")
+
+    # allow setting reservation hint as mandatory
+    reservation_required = CONF['blazar:physical:host'].blazar_reservation_required
+
+    reservation_hints = request_spec.scheduler_hints.get('reservation',None)
+
+    if reservation_required and not reservation_hints:
+        LOG.warning('Request did not specify mandatory reservation hint')
+        raise exception.RequestFilterFailed(
```
Comment (Mark-Powers) [inline nova/scheduler/request_filter.py]: Does this get exposed to the user?

Comment (msherman64) [reply nova/scheduler/request_filter.py]: The LOG.warning does not, but the message content of the exception does, in the `fault` field of the nova instance.

```json
{
  "fault": {
    "code": 500,
    "created": "2024-04-15T21:19:55Z",
    "message": "No valid host was found. Scheduling failed: No hosts available for tenant",
    "details": "Traceback (most recent call last):\n  File \"/var/lib.....<redacted for gh>"
  }
}
```

**Pull request closed without merging**